### DISCUSSION TITLE
TCP Skill vs TCPConnection

### DISCUSSION THREAD

**Discussion by hqnghi88** [gama-platform/gama#84] [Q&A]
Hi, 
In the recent commit (@lesquoyb  ), there is a new development of RawNetworkSkill and TCP skill. I am interested in the idea but how is different of this new type with TCP, UDP or MQTT, even with WebSocket in developing. Should we generalize/standardize a bit all these things (TCP UDP MQTT are strutctured by Nicolas and it seems perfect)

**Comment by lesquoyb** [comment]
Hello,
RawNetworkSkill is more oriented toward developers, it gives access to the "standard" tcp functions that you can have in most programming languages. The point was to get rid of the dependency from mailbox that is currently used with the network skill so I could use blocking functions in a simulation (which was not possible) and also have more control on the data exchanged.
I'm all for a generalization/standardization of the network functions, especially if we can reuse the same basic methods in the Gama software code (there are some parts that make http requests for example) and in agent skills, or even operators. Right now it would take me too much time to do it and I prefer working on the coming 1.8.2, but if someone has enough knowledge of those plugins and feels like doing it why not

**Comment by AlexisDrogoul** [comment]
Hi -- I think it would be clearer to have only one network skill, which could then offer "blocking" and "non-blocking" variants of the different commands (actually, what in the use of mailboxes prevent the commands from blocking a simulation 🤔 ?

**Reply by hqnghi88** [reply]
Yes, same question, can you specify more about the problem . Technically what you implement  (tcp socket) is already exist in TCPserver of network skill. Or you would like to have your own method for developing new feature, purely developers, it will not be a skill which is visible for modelers (non computer scientist.)

**Reply by lesquoyb** [reply]
I do not understand all the code from the network skill, but from what I can get it is very oriented toward a precise use with a lot of implicit things going on

Here is an example I have a really simple python server that is just meant to read data from a gama agent with network skill:
```
import socket
from _thread import *


def server_loop(sock_listen: socket) -> None:

    while True:
        conn, addr = sock_listen.accept()
        print("connected", conn, addr)
        while True:
            data = conn.recv(10240)
            print("python received", data)
        sock.close()


def listener_init() -> int:

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    print("Listening Socket successfully created")

    s.bind(('', 0))
    port = s.getsockname()[1]
    print("Listening socket binded to %s" % (port))

    s.listen()
    print("Listening socket started listening")

    start_new_thread(server_loop, (s,))

    return port


if __name__ == '__main__':

    #init server
    listener_port = listener_init()
    input()
```

And here is the gama model that is supposed to just send ONE message :

```
model NetworkingTest


global{
	
	init {
		do create_client;
	}

	action create_client {
		create Client number: 1 {
			do connect to: "localhost" protocol: "tcp_client" port: 56635 ;
		}
	}
	
}


species Client skills: [network]  parallel:true {
	
	reflex send {
		do send contents: name + " at " + cycle + " sent a message";
	}

}


experiment "Networking TCP Test" {
	
	init {
		create simulation;
	}
	
}
```

When running just one step of the simulation,  the expected output would be that the python code prints `python received b''client0 at 0 sent a message`. But instead I get this:

```
python received b'\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@null\n\r\n'
python received b'\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@null\n\r\n\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@ALL\n\r\n\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@ALL\n\r\n'
python received b'\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@null\n\r\n'
python received b'\x03\x05@b@@r@\x03\x05@b@@r@/127.0.0.1:56635\x03\x05@b@@r@localhost/127.0.0.1:56635\x03\x05@b@@r@0\x03\x05@b@@r@ALL\n\r\n'
python received b'\x03\x05@b@@r@null\x03\x05@b@@r@Client0\x03\x05@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <receivers class="msi.gama.metamodel.agent.MinimalAgent">@n@    <index>0</index>@n@    <variables>@n@      <keys>@n@        <string>name</string>@n@        <string>location</string>@n@        <string>shape</string>@n@        <string>mailbox</string>@n@        <string>network_name</string>@n@        <string>network_server</string>@n@        <string>network_groups</string>@n@      </keys>@n@      <data>@n@        <string>Client0</string>@n@        <msi.gama.metamodel.shape.GamaPoint>@n@          <x>47.36303990482106</x>@n@          <y>73.54858686057342</y>@n@          <z>0.0</z>@n@        </msi.gama.metamodel.shape.GamaPoint>@n@        <msi.gama.metamodel.shape.GamaShape>@n@          <geometry class="org.locationtech.jts.geom.Point">@n@            <factory class="msi.gama.common.geometry.GamaGeometryFactory">@n@              <precisionModel>@n@                <modelType>@n@                  <name>FLOATING</name>@n@                </modelType>@n@                <scale>0.0</scale>@n@              </precisionModel>@n@              <coordinateSequenceFactory class="msi.gama.common.geometry.GamaCoordinateSequenceFactory"/>@n@              <SRID>0</SRID>@n@            </factory>@n@            <SRID>0</SRID>@n@            <coordinates class="msi.gama.common.geometry.UniqueCoordinateSequence">@n@              <point reference="../../../../msi.gama.metamodel.shape.GamaPoint"/>@n@              <dimension>3</dimension>@n@            </coordinates>@n@          </geometry>@n@          <attributes class="msi.gama.util.GamaMap">@n@            <keysType class="msi.gaml.types.GamaStringType">@n@              <GamaType>string</GamaType>@n@            </keysType>@n@            <dataType class="msi.gaml.types.GamaNoType">@n@              <GamaType>unknown</GamaType>@n@            </dataType>@n@            <valuesMapReducer>@n@              <entry>@n@                <string>messaging_skill_mailbox</string>@n@                <msi.gama.extensions.messaging.GamaMailbox>@n@                  <valuesListReducer/>@n@                  <contentTypeListReducer class="msi.gaml.types.GamaMessageType">@n@                    <GamaType>message</GamaType>@n@                  </contentTypeListReducer>@n@                </msi.gama.extensions.messaging.GamaMailbox>@n@              </entry>@n@            </valuesMapReducer>@n@          </attributes>@n@        </msi.gama.metamodel.shape.GamaShape>@n@        <msi.gama.extensions.messaging.GamaMailbox reference="../msi.gama.metamodel.shape.GamaShape/attributes/valuesMapReducer/entry/msi.gama.extensions.messaging.GamaMailbox"/>@n@        <null/>@n@        <msi.gama.util.GamaList>@n@          <valuesListReducer>@n@            <string>localhost@@56635</string>@n@          </valuesListReducer>@n@          <contentTypeListReducer class="msi.gaml.types.GamaNoType" reference="../../msi.gama.metamodel.shape.GamaShape/attributes/dataType"/>@n@        </msi.gama.util.GamaList>@n@        <msi.gama.util.GamaList>@n@          <valuesListReducer>@n@            <string>ALL</string>@n@          </valuesListReducer>@n@          <contentTypeListReducer class="msi.gaml.types.GamaNoType" reference="../../msi.gama.metamodel.shape.GamaShape/attributes/dataType"/>@n@        </msi.gama.util.GamaList>@n@      </data>@n@    </variables>@n@  </receivers>@n@  <contents class="string">Client0 at 0 sent a message</contents>@n@  <emissionTimeStamp>0</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>\n\r\n'
python received b'\x03\x05@b@@r@null\x03\x05@b@@r@Client0\x03\x05@b@@r@<ummisco.gama.network.common.CompositeGamaMessage>@n@  <unread>true</unread>@n@  <receivers class="msi.gama.metamodel.agent.MinimalAgent">@n@    <index>0</index>@n@    <variables>@n@      <keys>@n@        <string>name</string>@n@        <string>location</string>@n@        <string>shape</string>@n@        <string>mailbox</string>@n@        <string>network_name</string>@n@        <string>network_server</string>@n@        <string>network_groups</string>@n@      </keys>@n@      <data>@n@        <string>Client0</string>@n@        <msi.gama.metamodel.shape.GamaPoint>@n@          <x>47.36303990482106</x>@n@          <y>73.54858686057342</y>@n@          <z>0.0</z>@n@        </msi.gama.metamodel.shape.GamaPoint>@n@        <msi.gama.metamodel.shape.GamaShape>@n@          <geometry class="org.locationtech.jts.geom.Point">@n@            <factory class="msi.gama.common.geometry.GamaGeometryFactory">@n@              <precisionModel>@n@                <modelType>@n@                  <name>FLOATING</name>@n@                </modelType>@n@                <scale>0.0</scale>@n@              </precisionModel>@n@              <coordinateSequenceFactory class="msi.gama.common.geometry.GamaCoordinateSequenceFactory"/>@n@              <SRID>0</SRID>@n@            </factory>@n@            <SRID>0</SRID>@n@            <coordinates class="msi.gama.common.geometry.UniqueCoordinateSequence">@n@              <point reference="../../../../msi.gama.metamodel.shape.GamaPoint"/>@n@              <dimension>3</dimension>@n@            </coordinates>@n@          </geometry>@n@          <attributes class="msi.gama.util.GamaMap">@n@            <keysType class="msi.gaml.types.GamaStringType">@n@              <GamaType>string</GamaType>@n@            </keysType>@n@            <dataType class="msi.gaml.types.GamaNoType">@n@              <GamaType>unknown</GamaType>@n@            </dataType>@n@            <valuesMapReducer>@n@              <entry>@n@                <string>messaging_skill_mailbox</string>@n@                <msi.gama.extensions.messaging.GamaMailbox>@n@                  <valuesListReducer/>@n@                  <contentTypeListReducer class="msi.gaml.types.GamaMessageType">@n@                    <GamaType>message</GamaType>@n@                  </contentTypeListReducer>@n@                </msi.gama.extensions.messaging.GamaMailbox>@n@              </entry>@n@            </valuesMapReducer>@n@          </attributes>@n@        </msi.gama.metamodel.shape.GamaShape>@n@        <msi.gama.extensions.messaging.GamaMailbox reference="../msi.gama.metamodel.shape.GamaShape/attributes/valuesMapReducer/entry/msi.gama.extensions.messaging.GamaMailbox"/>@n@        <null/>@n@        <msi.gama.util.GamaList>@n@          <valuesListReducer>@n@            <string>localhost@@56635</string>@n@          </valuesListReducer>@n@          <contentTypeListReducer class="msi.gaml.types.GamaNoType" reference="../../msi.gama.metamodel.shape.GamaShape/attributes/dataType"/>@n@        </msi.gama.util.GamaList>@n@        <msi.gama.util.GamaList>@n@          <valuesListReducer>@n@            <string>ALL</string>@n@          </valuesListReducer>@n@          <contentTypeListReducer class="msi.gaml.types.GamaNoType" reference="../../msi.gama.metamodel.shape.GamaShape/attributes/dataType"/>@n@        </msi.gama.util.GamaList>@n@      </data>@n@    </variables>@n@  </receivers>@n@  <contents class="string">Client0 at 0 sent a message</contents>@n@  <emissionTimeStamp>0</emissionTimeStamp>@n@</ummisco.gama.network.common.CompositeGamaMessage>\n\r\n'
```

So not only are there more messages sent than expected, but even the message is wrapped in an xml representing a GamaMessage.
I get that it is probably very useful in the context in which the skill was created but it is not easy to deal with this if you just wanted to send a string (and of course it is not efficient).

For the mailbox problem, I couldn't manage to "wait" for the mailbox to receive a message at a given step of the simulation, it caused an infinite loop. I tried to implement my methods in this skill aside from what was existing, but the way it works inside is completely different so I thought that creating a separate one was the best option for the user as well as for maintenance.

If you see a simple way to merge, them of course it would be nice, but from my point of view they are too different

**Reply by hqnghi88** [reply]
yes, because it is a composite message. It could be "plain" or "command" without any wrapping, adjusting information to the message. This can be provide in few hours, if you accept.

**Comment by AlexisDrogoul** [comment]
Messaging and serialisation are effectively implicit. But I would be more comfortable if we had only one set of underlying functions / libraries used to send/receive information on the network. The merge should be organised on these low-level function in Java. Above it, we can then think of having "raw" actions in addition to the "rich" ones, either identified by a different name or (I would prefer) resulting from the addition of a new facet in the existing commands (for instance `raw: true/false`) => when true, the execution is derived to the function you already coded (provided they use the same libraries) / when false, the default, we keep the same behaviour. 

What do you think ?

**Reply by hqnghi88** [reply]
I do agree. In fact, when i implement TCP and UDP, it was supposed to be as simple as @lesquoyb requirement, now. But then we moved to generalize/standardize the GAMA Message, and it become a more complex object. Anyway, it is already there some trace in network plugin (Type of message : command, plain, composite) which is should take into account in GAMA message: send "abc" type:command or plain.

**Comment by hqnghi88** [comment]
@lesquoyb 
I push the solution that Alexis said, please add a facet  raw: true to your connect statement. 
![image](https://user-images.githubusercontent.com/6105384/159512347-8a1a0a1a-28f4-471b-8d3c-8371b0eb11b2.png)
The message you will receive is raw then.
![image](https://user-images.githubusercontent.com/6105384/159512433-a2806e9f-55fb-4ccf-9ba0-42583459142f.png)

**Reply by lesquoyb** [reply]
Tested and it works well. It's not really a problem but is it possible to not end the sent messages with `\n\r\n` ?

**Reply by hqnghi88** [reply]
Yah  i know, but no time to see where is the \r\n , it must be in some serialize or toString, somewhere.

**Reply by benoitgaudou** [reply]
@hqnghi88 that's great we can send without the  whole GAMA Messsage!
It could be good  to implement it for all the protocols if  it is not the case yet, thanks  !

**Comment by AlexisDrogoul** [comment]
Great. Is there a correspondance between the sets of commands ? 
Now, there is the second problem that @lesquoyb was mentioning, concerning the possibility to block an agent while it is waiting for an answer or a message (something that already exists in FIPA using one of the high-level protocols). Remember that the point is not to block all GAMA systematically, but only one agent (and if this agent is a simulation, then all its agents are blocked of course) waiting for a network communication.

**Comment by hqnghi88** [comment]
You mean the subscribe, unsub and mailbox? it just bypass all that thing as "raw" message (but yes if need, i can take a look). But for "blocking" , i am a bit unsure. We can let agents do or not by checking messagebox is empty? it is not the case?

**Reply by lesquoyb** [reply]
The problem with just checking the messagebox is that you just check if you have a message at a given moment, while the functionality I was looking for is waiting for a message to come. For example an agent could need to wait for some server to send data in order to decide what to do in a reflex

**Reply by hqnghi88** [reply]
OK (not sure 100%, what to do in a reflex or even scheduling the agents can be decide with messagebox),  we can discuss more if you want, please share with me your idea in a simple model,

**Reply by lesquoyb** [reply]
For example let's say you want to have a moving agent that rely on a server for choosing it's direction, you could need to write something like this:
```
global{
	
	init {
		create Client number: 1 {
			do connect to: "localhost" protocol: "tcp_client" port: 57373 raw:true ;
		}
	}


}


species Client skills: [network, moving] {
	
	reflex guided_movement {
		do send contents: "I'm here: " + location;
		let orientation <- float(wait_message());
		move heading:orientation ;
	}

}

```

At every step the agent sends its location to a server, the server do some processing and then gives back to the agent the correct heading to reach a goal, only then the agent can use the move function.

I just had a chat with @AlexisDrogoul about how we should "wait". What I did in TcpSkill was to block the whole thread, which for my use was fine, but would not work in other cases. So we must find a way to "stop" only the agent waiting and the rest of the simulation should continue to run

**Reply by hqnghi88** [reply]
Hi Baptiste, what is the real use case? something like online game scenario ?
for the example, i dont see why something like this (in the example) did not mean "blocking" your using of agents
![image](https://user-images.githubusercontent.com/6105384/159684037-663457fe-f70e-40e1-89ca-367ccfc5d2f7.png)
or even scheduling
![image](https://user-images.githubusercontent.com/6105384/159684472-753d8810-0304-4718-b11a-f6300b6b3519.png)
Anyway, if you have discussed with Alexis, we should find a way to "stop" for your case.

**Reply by lesquoyb** [reply]
Hi nghi,
The real use case is a gama agent that is supposed to pick up policies to optimize certain aspects of a simulation. We are relying on a python neural net for processing which policy to choose, so we figured out that the best way to do it was just to have communication between the python script and the gama agent via tcp.
The problem I had with the code you posted is that the reflex would only trigger at least one cycle after the message is actually needed. Now with this code placed at the simulation level and with our particular use case we could make some workaround even though I still find it unpractical. And it would require for the modeler to really take into account that maybe it will take an unknown number of cycles for the simulation to complete one "real step"

**Reply by hqnghi88** [reply]
OK, in that case, i do agree with you that we need special technic to facilitate the implementation for modeler. As your need is an operation smaller than the atom "step" , i think the "raw" facet still feasible but need more changes in the way mailbox are listening. I will try something and report back.

**Reply by hqnghi88** [reply]
[TCP Server And Client Example 2.txt](https://github.com/gama-platform/gama/files/8343229/TCP.Server.And.Client.Example.2.txt)
Here is an example that i tried to play with current syntax, if you find it fit with the use case, we can try to implement in java level. Or you can  suggest the missing idea.
For the moment, there is the bug of xstream that make some error when execute network skill, but we will soon fix it. 
Other thing, the synchronization of the display of the agents will do 1 times, so at a moment, if agent A was blocking , it received message and moving, the visual effect will not be updated until next redraw. @AlexisDrogoul do you have any idea about flexible displaying/updating/redraw of a set of agents in a specific step?

**Reply by lesquoyb** [reply]
Yep that could be a workaround for my specific use case. But right now I tested it with a gama simulation and python script and it doesn't seem to work, the messages sent by python never seem to be read in the simulation. Also I [commited a small fix](https://github.com/gama-platform/gama/commit/5503a9ea6e0e835951ec21ec092c95bf8491eaa3) related to receiving small messages.

I'm not sure what you meant with the display issue, can you elaborate more ? Do you mean that we would have drawing issues with blocking agents ?

**Comment by RoiArthurB** [comment]
FYI : Your work did broke regression tests on [automatic build](https://github.com/gama-platform/gama/runs/5657622198?check_suite_focus=true#step:6:113) : 
```
----------------------------------------------------------------
failed: Tests for ummisco.gama.network in network_test_interaction 
----------------------------------------------------------------
passed: Test test Message sending 
passed: test 

passed: Test test Message receiving 
passed: test 

passed: Test test Message contents 
passed: test 

passed: Test test has_more_element and fetch_message action 
passed: ((a1 and a2) and a3) and a4 

failed: Test test network group - broadcast to ALL 
failed: ((p1 and p2) and p3) and p4 

failed: Test test network group - broadcast to group Spice girl 
failed: p1 
```

**Reply by RoiArthurB** [reply]
The first test failed comes from : `ummisco.gama.network/tests/Network Tests/models/network_test_interaction.experiment`

And the second from : `ummisco.gama.network/tests/Network Tests/models/network_test_interaction.experiment`

**Reply by hqnghi88** [reply]
euh, Spice girl still accept my message 
![image](https://user-images.githubusercontent.com/6105384/159668847-47ecf2c2-7d43-4b49-b005-8301a0b423a9.png)

**Reply by RoiArthurB** [reply]
The local version works, it's the not-local which failed : 
```
[...]

2022-03-23T09:19:30.3478156Z network group -- done!
2022-03-23T09:19:30.3540017Z reportAndThrowIfNeeded : Assert failed: ((p1 and p2) and p3) and p4
2022-03-23T09:19:33.8227901Z network group Spice Girls -- done!
2022-03-23T09:19:33.8233054Z reportAndThrowIfNeeded : Assert failed: p1
2022-03-23T09:19:33.8283068Z ----------------------------------------------------------------
2022-03-23T09:19:33.8283594Z failed: Tests for ummisco.gama.network in network_test_interaction 
2022-03-23T09:19:33.8284719Z ----------------------------------------------------------------
2022-03-23T09:19:33.8285238Z passed: Test test Message sending 
2022-03-23T09:19:33.8285568Z passed: test 
2022-03-23T09:19:33.8285774Z 
2022-03-23T09:19:33.8285959Z passed: Test test Message receiving 
2022-03-23T09:19:33.8287166Z passed: test 
2022-03-23T09:19:33.8287391Z 
2022-03-23T09:19:33.8287587Z passed: Test test Message contents 
2022-03-23T09:19:33.8287903Z passed: test 
2022-03-23T09:19:33.8288283Z 
2022-03-23T09:19:33.8288538Z passed: Test test has_more_element and fetch_message action 
2022-03-23T09:19:33.8288913Z passed: ((a1 and a2) and a3) and a4 
2022-03-23T09:19:33.8289135Z 
2022-03-23T09:19:33.8289634Z failed: Test test network group - broadcast to ALL 
2022-03-23T09:19:33.8290163Z failed: ((p1 and p2) and p3) and p4 
2022-03-23T09:19:33.8290377Z 
2022-03-23T09:19:33.8290674Z failed: Test test network group - broadcast to group Spice girl 
2022-03-23T09:19:33.8291014Z failed: p1 
2022-03-23T09:19:33.8291205Z 
2022-03-23T09:19:33.8291572Z ----------------------------------------------------------------
2022-03-23T09:19:33.8291994Z failed: Tests for ummisco.gama.network in network_test_interaction 
2022-03-23T09:19:33.8292477Z ----------------------------------------------------------------
2022-03-23T09:19:33.8292846Z passed: Test test Message sending 
2022-03-23T09:19:33.8293145Z passed: test 
2022-03-23T09:19:33.8293335Z 
2022-03-23T09:19:33.8293507Z passed: Test test Message receiving 
2022-03-23T09:19:33.8293812Z passed: test 
2022-03-23T09:19:33.8294002Z 
2022-03-23T09:19:33.8294178Z passed: Test test Message contents 
2022-03-23T09:19:33.8294483Z passed: test 
2022-03-23T09:19:33.8294675Z 
2022-03-23T09:19:33.8294884Z passed: Test test has_more_element and fetch_message action 
2022-03-23T09:19:33.8295762Z passed: ((a1 and a2) and a3) and a4 
2022-03-23T09:19:33.8296165Z 
2022-03-23T09:19:33.8296485Z failed: Test test network group - broadcast to ALL 
2022-03-23T09:19:33.8296837Z failed: ((p1 and p2) and p3) and p4 
2022-03-23T09:19:33.8297060Z 
2022-03-23T09:19:33.8298737Z failed: Test test network group - broadcast to group Spice girl 
2022-03-23T09:19:33.8299159Z failed: p1 
2022-03-23T09:19:33.8299535Z 
2022-03-23T09:19:36.0388544Z Message sending -- done!
2022-03-23T09:19:37.0398322Z Message receiving -- done!
2022-03-23T09:19:38.0406515Z Message contents -- done!
2022-03-23T09:19:40.0438266Z has_more_element and fetch_message action -- done!
2022-03-23T09:19:42.1653982Z network group -- done!
2022-03-23T09:19:45.9297612Z network group Spice Girls -- done!
2022-03-23T09:19:45.9319802Z ----------------------------------------------------------------
2022-03-23T09:19:45.9322959Z passed: Tests for ummisco.gama.network in network_test_interaction_local 
2022-03-23T09:19:45.9324076Z ----------------------------------------------------------------
2022-03-23T09:19:45.9527646Z passed: Test test Message sending 
[...]
```

**Reply by hqnghi88** [reply]
the same, Girls are still on my site
![image](https://user-images.githubusercontent.com/6105384/159669883-932d3cb1-bcd6-41d8-9021-36253354b9b6.png)

**Reply by hqnghi88** [reply]
and the latest commit 
![image](https://user-images.githubusercontent.com/6105384/159670330-c0dbc174-2a40-43e2-8c71-9650906dbe4c.png)

**Reply by RoiArthurB** [reply]
Just restart a built, and now everything works 🤷🤷🤷

```
GAMA test is starting...
******************************************************************
* GAMA version 1.8.2                                             *
* http://gama-platform.org/                                       *
* (c) 2007-2021 UMI 209 UMMISCO IRD/SU & Partners                *
******************************************************************
> GAMA: version 1.8.2                         loading on____ linux 5.11.0-1028-azure, x86_64, JDK 17.0.2
> JAI : ImageIO extensions                    loaded for____ |jpg|tiff|bmp|gif|arx|tf8|TF8|png|ppm|jp2|tif|TIF|asc|TIFF|btf|BTF|pgm|wbmp|jpeg|pbm
> GAMA: msi.gama.core                         loaded in_____ 1460ms
> GAMA: ummisco.gama.serialize                loaded in_____ 5ms
> GAMA: ummisco.gama.network                  loaded in_____ 8ms
> GAMA: ummisco.gama.opengl                   loaded in_____ 6ms
> GAMA: msi.gaml.architecture.simplebdi       loaded in_____ 70ms
> GAMA: ummisco.gaml.extensions.maths         loaded in_____ 84ms
> GAMA: ummisco.gama.java2d                   loaded in_____ 2ms
> GAMA: msi.gama.headless                     loaded in_____ 2ms
> GAMA: msi.gama.lang.gaml                    loaded in_____ 3ms
> GAMA: msi.gaml.extensions.fipa              loaded in_____ 12ms
> GAMA: espacedev.gaml.extensions.genstar     loaded in_____ 6ms
> GAMA: miat.gaml.extension.pedestrian        loaded in_____ 17ms
> GAMA: ummisco.gaml.extensions.stats         loaded in_____ 24ms
> GAMA: simtools.gaml.extensions.traffic      loaded in_____ 38ms
> GAMA: simtools.gaml.extensions.physics      loaded in_____ 11ms
> GAMA: irit.gaml.extensions.database         loaded in_____ 13ms
> GAMA: all plugins                           loaded in_____ 1854ms
GAMA configuring and loading...
> GAMA: GAML artefacts                        built in______ 57ms
Message sending -- done!
Message receiving -- done!
Message contents -- done!
has_more_element and fetch_message action -- done!
network group -- done!
network group Spice Girls -- done!
----------------------------------------------------------------
passed: Tests for ummisco.gama.network in network_test_interaction 
----------------------------------------------------------------
passed: Test test Message sending 
passed: test 
passed: Test test Message receiving 
passed: test 
passed: Test test Message contents 
passed: test 
passed: Test test has_more_element and fetch_message action 
passed: ((a1 and a2) and a3) and a4 
passed: Test test network group - broadcast to ALL 
passed: ((p1 and p2) and p3) and p4 
passed: Test test network group - broadcast to group Spice girl 
passed: p1 
Message sending -- done!
Message receiving -- done!
Message contents -- done!
has_more_element and fetch_message action -- done!
network group -- done!
network group Spice Girls -- done!
----------------------------------------------------------------
passed: Tests for ummisco.gama.network in network_test_interaction_local 
----------------------------------------------------------------
passed: Test test Message sending 
passed: test 
passed: Test test Message receiving 
passed: test 
passed: Test test Message contents 
passed: test 
passed: Test test has_more_element and fetch_message action 
passed: ((a1 and a2) and a3) and a4 
passed: Test test network group - broadcast to ALL 
passed: ((p1 and p2) and p3) and p4 
passed: Test test network group - broadcast to group Spice girl 
passed: p1
[...]
```

**Comment by hqnghi88** [comment]
you had to restart even that the latest passed well???
